### ISSUE TITLE
Endpoint `explode` does not destroy the player ship if it is strong enough.

### ISSUE THREAD

**Issue by Pshy0** [endless-sky/endless-sky#8831]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

A strong enough ship will not be destroyed by the `explode` endpoint, and will only be disabled.
The player apparently is considered dead anyway as it's fleet goes away.
But the player can get help from another ship and get fixed, but will be considered dead as soon as they land.

### Steps to Reproduce

1. Get your flagship to be a `Rano'erek` (or something stronger, it also works with a 577).
2. Run a mission that run the explode endpoint

The mission:
```
mission "Explode"
	landing
	repeat
	on enter
		conversation
			`Explode now?`
			choice
				`	(explode)`
					explode
				`	(not now)`
					defer
```

### Expected Behavior

The player's flagship should explode.

### Screenshots

![image](https://github.com/endless-sky/endless-sky/assets/44928041/fb0161ac-afcf-40ff-8714-9edec44d8cde)

### Link to save file

The ship:
```
ship "Korath Dredger"
	name Adwaska
	sprite ship/dredger
	thumbnail thumbnail/dredger
	uuid b6a060a3-f0d2-4acf-adfa-28ef348da243
	attributes
		category "Heavy Freighter"
		cost 29370000
		mass 1503
		bunks 395
		"cargo space" 605
		drag 18.3
		"engine capacity" 181
		"fuel capacity" 900
		"heat dissipation" 0.65
		hull 16000
		"outfit space" 827
		"required crew" 201
		shields 39000
		"turret mounts" 7
		"weapon capacity" 297
	outfits
		`"Bufaer" Atomic Steering`
		`"Bufaer" Atomic Thruster`
		"Antimatter Core"
		"Bunk Room" 2
		"Cloaking Device"
		"Control Transceiver"
		"Cooling Module"
		"Enforcer Confrontation Gear" 697
		"Enforcer Riot Staff" 1057
		"Fragmentation Grenades" 400
		"Intrusion Countermeasures" 640
		"Jump Drive"
		"Korath Repeater Rifle" 772
		"Laser Rifle" 2
		"Lasher Pistol"
		"Nanotech Battery" 2
		"Outfits Expansion" 30
		"Overcharged Shield Module" 12
		"Overclocked Repair Module"
		"Pug Biodefenses" 164
		"Pug Peacekeeping Staff" 570
		"Pulse Rifle" 400
		"Quantum Keystone"
		"Quarg Anti-Missile" 2
		"Quarg Skylance" 2
		"Research Laboratory"
		"Salvage Scanner"
		"Scanning Module"
		"Tuning Rifle" 10
		"Wanderer Ramscoop" 8
	crew 204
	fuel 900
	shields 39000
	hull 16000
	position 252.2775 -249.22483
	engine 0 179
		zoom 1
		angle 0
		under
	engine -32 164
		zoom 0.7
		angle 0
		under
	engine 32 164
		zoom 0.7
		angle 0
		under
	"reverse engine" -93.5 -65
		zoom 1
		angle -360
		under
	"reverse engine" 93.5 -65
		zoom 1
		angle -360
		under
	"reverse engine" -79 80.5
		zoom 1
		angle -360
		over
	"reverse engine" 79 80.5
		zoom 1
		angle -360
		over
	turret -37 -155 "Quarg Skylance"
		over
	turret 0 -155 "Quarg Skylance"
		over
	turret 37 -155 "Quarg Anti-Missile"
		over
	turret -96 0 "Quarg Anti-Missile"
		over
	turret 96 0
		over
	turret -80 130
		over
	turret 80 130
		over
	bay Fighter -93.5 -20.5
		"launch effect" "basic launch"
	bay Fighter 93.5 -20.5
		"launch effect" "basic launch"
	bay Fighter -87 128.5
		"launch effect" "basic launch"
	bay Fighter 87 128.5
		"launch effect" "basic launch"
	explode "huge explosion" 15
	explode "large explosion" 50
	explode "medium explosion" 70
	explode "small explosion" 60
	explode "tiny explosion" 120
	"final explode" "final explosion large" 1
	system Cinxia
	planet Caelian
```

### Operating System

Ubuntu

### Game Source

GitHub Releases

### Game Version

v0.10.1

**Comment by TomGoodIdea** [comment]
I can confirm this and I suspect it's a floating-point calculation error.

**Comment by TomGoodIdea** [comment]
`Ship::Destroy()` sets hull value to `-1.` and then `Ship::IsDestroyed()` checks if the value is lower than `0.`, but somehow the check returns false for larger ships. Lowering the value in the Destroy function fixes the bug, but I'm not sure if that's the best solution.

**Comment by EjoThims** [comment]
Have/can you check with a smaller ship that has high hull regen and a larger ship that has zero hull regen?

Based on your description, it sounds like it may be because a single frame of hull regen is bringing the ship back above 0.??

**Comment by tehhowch** [comment]
> `Ship::Destroy()` sets hull value to `-1.` and then `Ship::IsDestroyed()` checks if the value is lower than `0.`, but somehow the check returns false for larger ships. Lowering the value in the Destroy function fixes the bug, but I'm not sure if that's the best solution.

Correct, the fix is to ensure that things occur in the correct order, or barring that, that hull regen (and other things) are not done if a ship is "already" dead.

**Comment by UnorderedSigh** [comment]
My fix for this is #8851, and follows @tehhowch's suggestion. To do this, I had to split Ship::Move into multiple methods.

**Comment by UnorderedSigh** [comment]
My fix for this is #8855, and follows @tehhowch's suggestion. To do this, I had to split Ship::Move into multiple methods.

Edit: My previous comment had the wrong PR number.
